### ISSUE TITLE
IQueryable<> does not utilize custom CosmosSerializer when serializing datetimeoffset

### ISSUE THREAD

**Issue by onionhammer** [Azure/azure-cosmos-dotnet-v3#3808]
**Describe the bug**
When serializing DateTimeOffset values, they are formatted like "2023-04-11T21:24:36.6331736+00:00" instead of like "2023-04-11T21:24:36.6331736Z", even though I have configured a custom json serializer which formats DateTimeOffset's properly for cosmos range indexing.

The result is that filters based on datetimeoffsets are not accurate

![image](https://user-images.githubusercontent.com/969938/231293256-3e1acad4-2975-4645-b972-65b23630dd2f.png)

**To Reproduce**
Steps to reproduce the behavior. If you can include code snippets or links to repositories containing a repro of the issue that can helps us in detecting the scenario it would speed up the resolution.

**Expected behavior**
`ToQueryDefinition()` should run values through the CosmosSerializer. This is also related to enums serialization issues (serializing as numbers).

**Actual behavior**
DateTimeOffset formatted using some newtonsoft json serializer defaults

**Environment summary**
SDK Version: 3.32.3
OS Version: Widows 10

**Comment by ealsur** [comment]
@onionhammer is this not related to the fact that LINQ does not correctly use the custom Serializer?

**Comment by onionhammer** [comment]
> This is also related to enums serialization issues (serializing as numbers).

Yes, that's what I meant here.. but shouldn't Cosmos SDK **by default** [serialize dates into the format specified by Cosmos docs](https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/query/working-with-dates#storing-datetimes)?

**Comment by onionhammer** [comment]
@ealsur I feel the resolution of that issue has stalled, and this could be resolved without that fix - but it would also be partially resolved BY that fix, but wouldn't change the default behavior, which is to serialize dates in a format Cosmos itself does not understand as a date.

**Comment by onionhammer** [comment]
@ealsur

> When you mention about serializing the date, the scope is still the query, correct

I'm referring to formatting the query, not when saving the data; in my use-case I'm not using the Newtonsoft serializer for saving/reading data, I'm using a custom serializer which uses STJ.

the code to create the query is like the following:

```C#
DateTimeOffset dateValue = DateTimeOffset.UtcNow.AddDays(-5);
Container.GetItemLinqQueryable<Snapshot>()
    .Where(c => c.DateLastEvent > dateValue)
    .ToQueryDefinition();
```

**Comment by ealsur** [comment]
@onionhammer The default serialization engine is Newtonsoft.Json. Newtonsoft.Json by default, serializes DateTimeOffset with the format: `"2023-04-07T15:27:33.415438+00:00"` (for your example of `DateTimeOffset.UtcNow.AddDays(-5)`). For example,

```
public static void Main()
{
	MyType myType = new MyType();
	myType.MyDateOffset =  DateTimeOffset.UtcNow.AddDays(-5);
	myType.MyDateTime =  DateTime.UtcNow.AddDays(-5);
	Console.WriteLine(Newtonsoft.Json.JsonConvert.SerializeObject(myType));
}

public class MyType
{
	public DateTimeOffset MyDateOffset {get;set;}
	public DateTime MyDateTime {get;set;}
}
```

Prints:

```
{"MyDateOffset":"2023-04-07T15:31:40.6588321+00:00","MyDateTime":"2023-04-07T15:31:40.6588321Z"}
```

This means that there is no issue in itself in terms of the default behavior, it still follows Newtonsoft.Json's behavior. Why does Newtonsoft.Json decide to serialize DatetimeOffset differently than DateTime that I don't know. You can however customize it with a Converter: https://stackoverflow.com/a/23505560/5641598

In your case though, you are using System.Text.Json through a Custom Serializer, and we know based on the other Issue that LINQ would not work as expected. 

My point is, if this issue is about DateTimeOffset serialization (default engine) then the answer is that the behavior is the expected one for Newtonsoft.Json and if you want a different one, you could use a Converter. If the issue is about LINQ and Custom Serializer, why not add it to the existing Issue instead of creating more?

**Comment by onionhammer** [comment]
@ealsur "2023-04-07T15:31:40.6588321+00:00" is not the format specified in the article I linked above. That's the whole issue.

**Comment by ealsur** [comment]
@onionhammer That article is not using `DateTimeOffset`, it is using `DateTime`, the article is valid and the code would behave the same way if `DateTime` is used. I am sorry, but I don't see where the article and the expectations mismatch if that type is used?

**Comment by onionhammer** [comment]
@ealsur "2023-04-07T15:31:40.6588321+00:00" is a DateTimeOffset, because it includes a time zone. A DateTime does not include an offset.

**Comment by ealsur** [comment]
> they both should result in the same format string when communicating with Cosmos.

I understand the expectation but that is not how Newtonsoft.Json works (for better or worse). Newtonsoft.Json serializes `DateTimeOffset` differently, that is how it is.

You can certainly customize the serialization for `DateTimeOffset` to make it like `DateTime`:

```
class UtcDateTimeOffsetConverter : Newtonsoft.Json.Converters.IsoDateTimeConverter
    {
        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
        {
            if (value is DateTimeOffset)
            {
                var date = (DateTimeOffset)value;
                value = date.UtcDateTime;
            }
            base.WriteJson(writer, value, serializer);
        }

        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
        {
            object value = base.ReadJson(reader, objectType, existingValue, serializer);
            if (value is DateTimeOffset)
            {
                var date = (DateTimeOffset)value;
                value = date.ToLocalTime();
            }
            return value;
        }
    }
```

```
DateTimeOffset dateValue = DateTimeOffset.UtcNow.AddDays(-5);
var qf = container.GetItemLinqQueryable<MyType>()
            .Where(c => c.DateLastEvent > dateValue)
            .ToQueryDefinition();
```

Now generates `SELECT VALUE root FROM root WHERE (root["DateLastEvent"] > "2023-04-07T16:08:58.0565997Z")`

**Comment by onionhammer** [comment]
That makes sense, thank you @ealsur

I do think there's a debate to be had as to whether or not that should be the default for LINQ queries (Newtonsoft defaults aside), but I am fine using a workaround in this instance.

I'm struggling to think of a case where someone would ever want to query Cosmos with `c.Date > "2023-04-07T15:27:33.415438+00:00"` since that would yield a string comparison rather than an date range comparison, so the result would be meaningless.

**Comment by ealsur** [comment]
I completely agree with the expectations in terms of the format, I too would expect both to have the same, but for whatever reason, Newtonsoft.Json decided to diverge on how it serializes one or the other. I believe that might be the reason the documentation article was using `DateTime` and it does not mention `DateTimeOffset` (I can only assume since I don't know who wrote it).

The scope of this issue seems resolved though, as the behavior (`DateTimeOffset` serialization) is clear, and there is a way to change it to the desired one (either that or use `DateTime`).

For using STJ or any other custom serializer with LINQ, I believe the other issue still would cover it.

**Comment by onionhammer** [comment]
I guess I was more raising the question of 'should Cosmos .NET SDK be reliant on how Newtonsoft Json happens to serialize DateTimeOffset by default even though it doesn't make sense for the particular use-case of Cosmos?', not really interested in 'How does Newtonsoft happen to serialize DateTimeOffset', but 'How should Cosmos .NET SDK serialize datetimeoffset'

**Comment by ealsur** [comment]
@onionhammer fair question. The answer is that the .NET SDK does rely on Newtonsoft.Json because that is the chosen serialization engine. Based on the article and documentation, the guideline would be to use `DateTime` when working with dates and expecting date comparison, and within those guidelines, the SDK works as expected.

We could internally treat `DateTimeOffset` differently but then the problem is that it would be a behavior change that might break existing customers, some people, for whatever reason, might have already depended on the SDK following Newtonsoft.Json's behavior (they might have built their own converters, modified Newtonsoft.Json configuration, or created UDFs for example, or they don't even do date comparison, so the current format works fine for them), I don't think this is something we can change. That is probably why the reason that the article guides customers into using `DateTime`, not `DateTimeOffset` for cases where you want to allow dates comparison.

By the way, System.Text.Json's default serialization behavior also matches Newtonsoft.Json for `DateTimeOffset`.

**Comment by onionhammer** [comment]
@ealsur valid concern re: backwards compatibility, but there needs to be some consideration for new implementers who, while implementing new applications using this SDK, may miss the subtle misbehavior between a string comparison and a date range comparison being done within Cosmos.

Perhaps this distinction should be added as a warning to the documentation, at the very least.

Many new .NET applications don't mess around with DateTime objects anymore, since DateTimeOffset avoids any ambiguities around DateTimeKind, so this would be an immediate foot-gun built into the SDK that could be very difficult to notice.

**Comment by ealsur** [comment]
> Perhaps this distinction should be added as a warning to the documentation, at the very least.

I agree, the documentation should be updated to reflect guidance when working with `DateTimeOffset` due to the context you describe.

@jcocchi Can we track improving this doc -> https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/query/working-with-dates#querying-datetimes-in-linq to account for users using `DateTimeOffset`? We could add references to an example code if needed (like https://github.com/Azure/azure-cosmos-dotnet-v3/issues/3808#issuecomment-1505551450)?

**Comment by onionhammer** [comment]
> Or is it also failing to work with `DateTime`?

It is also failing with datetime, I think

**Comment by ealsur** [comment]
> It is also failing with datetime, I think

Do you have a repro? When I tested DateTime it worked fine, for DateTimeOffset, I had to add the converter. Above [comment](https://github.com/Azure/azure-cosmos-dotnet-v3/issues/3808#issuecomment-1505518183) shows the serialization of DateTime.

**Comment by onionhammer** [comment]
> > It is also failing with datetime, I think
> 
> Do you have a repro? When I tested DateTime it worked fine, for DateTimeOffset, I had to add the converter. Above [comment](https://github.com/Azure/azure-cosmos-dotnet-v3/issues/3808#issuecomment-1505518183) shows the serialization of DateTime.

It's when comparing against a DateTimeOffset

![image](https://user-images.githubusercontent.com/969938/236286045-d892a683-f072-4e13-9ff1-5850cb0644d8.png)

**Comment by ealsur** [comment]
@onionhammer Thanks for confirming DateTime is not a problem, then this is still the same topic we were discussing earlier and linked to the documentation update.

**Comment by onionhammer** [comment]
> @onionhammer Thanks for confirming DateTime is not a problem, then this is still the same topic we were discussing earlier and linked to the documentation update.

It is only a problem when comparing it with a DateTimeOffset - in the screenshot it is using the wrong format

**Comment by ealsur** [comment]
Right, so the documentation should give users accordingly, this is what we are tracking with this issue. The root is that Newtonsoft serializes these two date types differently, there are no code changes that can be done on the SDK because they will be a breaking change, we already established that on this thread.

**Comment by onionhammer** [comment]
@Maya-Painter is this scenario fixed by #4138  ?

**Comment by Maya-Painter** [comment]
@onionhammer can I close this issue as resolved in the newest release? 

Reference #4298

**Comment by onionhammer** [comment]
@Maya-Painter yep, didn't realize this one was still open

Closed with #4298
